cluster-autoscaler/utils/kubernetes/listers.go

```diff
@@ -153,67 +152,51 @@ type PodLister interface {
 	List() ([]*apiv1.Pod, error)
 }
 
-// UnschedulablePodLister lists unscheduled pods
-type UnschedulablePodLister struct {
+// ScheduledPodLister lists scheduled pods.
+type ScheduledPodLister struct {
 	podLister v1lister.PodLister
 }
 
-// List returns all unscheduled pods.
-func (unschedulablePodLister *UnschedulablePodLister) List() ([]*apiv1.Pod, error) {
-	var unschedulablePods []*apiv1.Pod
-	allPods, err := unschedulablePodLister.podLister.List(labels.Everything())
-	if err != nil {
-		return unschedulablePods, err
-	}
-	for _, pod := range allPods {
-		_, condition := podv1.GetPodCondition(&pod.Status, apiv1.PodScheduled)
-		if condition != nil && condition.Status == apiv1.ConditionFalse && condition.Reason == apiv1.PodReasonUnschedulable {
-			unschedulablePods = append(unschedulablePods, pod)
-		}
-	}
-	return unschedulablePods, nil
-}
-
-// NewUnschedulablePodLister returns a lister providing pods that failed to be scheduled.
-func NewUnschedulablePodLister(kubeClient client.Interface, stopchannel <-chan struct{}) PodLister {
-	return NewUnschedulablePodInNamespaceLister(kubeClient, apiv1.NamespaceAll, stopchannel)
+// List returns all scheduled pods.
+func (lister *ScheduledPodLister) List() ([]*apiv1.Pod, error) {
+	return lister.podLister.List(labels.Everything())
 }
 
-// NewUnschedulablePodInNamespaceLister returns a lister providing pods that failed to be scheduled in the given namespace.
-func NewUnschedulablePodInNamespaceLister(kubeClient client.Interface, namespace string, stopchannel <-chan struct{}) PodLister {
+// NewScheduledPodLister builds ScheduledPodLister
+func NewScheduledPodLister(kubeClient client.Interface, stopchannel <-chan struct{}) PodLister {
 	// watch unscheduled pods
-	selector := fields.ParseSelectorOrDie("spec.nodeName==" + "" + ",status.phase!=" +
+	selector := fields.ParseSelectorOrDie("spec.nodeName!=" + "" + ",status.phase!=" +
 		string(apiv1.PodSucceeded) + ",status.phase!=" + string(apiv1.PodFailed))
-	podListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", namespace, selector)
+	podListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", apiv1.NamespaceAll, selector)
 	store, reflector := cache.NewNamespaceKeyedIndexerAndReflector(podListWatch, &apiv1.Pod{}, time.Hour)
 	podLister := v1lister.NewPodLister(store)
 	go reflector.Run(stopchannel)
-	return &UnschedulablePodLister{
+
+	return &ScheduledPodLister{
 		podLister: podLister,
 	}
 }
 
-// ScheduledPodLister lists scheduled pods.
-type ScheduledPodLister struct {
+// ScheduledAndUnschedulablePodLister lists scheduled and unschedulable pods.
+type ScheduledAndUnschedulablePodLister struct {
 	podLister v1lister.PodLister
 }
 
-// List returns all scheduled pods.
-func (lister *ScheduledPodLister) List() ([]*apiv1.Pod, error) {
+// List returns all scheduled and unschedulable pods.
+func (lister *ScheduledAndUnschedulablePodLister) List() ([]*apiv1.Pod, error) {
```

Comment:
This is a major change in the logic. "Unschedulable" pods are pods that the scheduler explicitly marks as such using the `PodScheduled` condition with `PodReasonUnschedulable`, not just pods that don't have a `nodeName` assigned. I can't recall particular examples, but I think the scheduler intentionally doesn't mark some pods as unschedulable (while leaving `nodeName` empty). Currently CA would not react to such pods, after these changes it would start to.

In particular in certain scaling scenarios we rely on scheduler's marking of pods as unschedulable as a rate limiting mechanism (if a lot of new pods are created at once, scheduler effectively rate-limits how many CA has to handle at once because it has to mark the pods first). We'd lose that ability after these changes.

IMO this method should list the pods from `podLister`, filter them into two lists:
* Pods with `nodeName != ""` -> scheduled pods
* Pods with `nodeName == ""` AND the Unschedulable condition present (you can use the old logic for that check) -> unschedulable pods

And then return both lists separately. This should keep the logic intact and improve responsibility separation (it should be the listers job to separate).

Reply:
Done